### ISSUE TITLE
Pass test Method to TestInstanceFactory when using Lifecycle.PER_METHOD

### ISSUE THREAD

**Issue by seanf** [junit-team/junit5#1568]
I'm trying to update [CDI-Unit](http://bryncooke.github.io/cdi-unit) to [support JUnit 5](https://github.com/BrynCooke/cdi-unit/issues/103), specifically CDI-Unit's [`ProducerConfig` feature for JUnit 4](https://github.com/BrynCooke/cdi-unit/blob/805e077660d7d609b24b0d2b8dc1fa264ad30082/cdi-unit/src/test/java/org/jglue/cdiunit/TestProducerConfig.java) which uses annotations on the test method to help configure the deployment payload for Weld. Weld is then used to construct test instances. This can almost be replicated in JUnit 5, by implementing a `TestInstanceFactory`.  (Obviously this won't be compatible with `Lifecycle.PER_CLASS`, because the instances must allow for per-method configuration.)

The problem is, when `TestInstanceFactory.createTestInstance()` is called, `extensionContext.getTestMethod()` always returns `Optional.empty()`, even when using `Lifecycle.PER_METHOD`.

If the `TestInstanceFactory` were to receive the test Method, when using `Lifecycle.PER_METHOD`, it would be possible to use annotations from the test method to control test instance creation, thus allowing features like `ProducerConfig` to be created.

**Comment by sbrannen** [comment]
FWIW, you may be interested in #878.

**Comment by sbrannen** [comment]
In any case, what I stated above is naturally based on the status quo.

It _may_ be technically possible to redesign the internals such that `TestMethodTestDescriptor.prepare(JupiterEngineExecutionContext)` passes the current test `Method` to `TestInstanceProvider.getTestInstance(Optional<ExtensionRegistry>)` to allow our internal implementations to create a one-off, modified `ExtensionContext` that contains the "method about to be executed".

_maybe_...

**Comment by seanf** [comment]
> That means that the underlying ExtensionContext is scoped at the class level. Specifically, it is a ClassExtensionContext which has no access to the Method.

Right, I see.

> Furthermore, doing so would not align with the design of JUnit Jupiter, namely that class-level extensions should not do anything particular to specific methods.

I'm not sure I understand what you mean by class-level extensions.

Perhaps the title of this feature request is too specific in its details (ie TestInstanceFactory), but it would still be useful IMHO, when using PER_METHOD Lifecycle, to be able construct test instances based on method attributes.

**Comment by sbrannen** [comment]
> I'm not sure I understand what you mean by class-level extensions.

Extensions fall into certain categories depending on where they are applied / when they are invoked.

For example, `BeforeEachCallback` is always a method-level extension; whereas, `BeforeAllCallback` is always a _container-level_ extension.

Specifically, `BeforeAllCallback` is invoked as a class-level extension if the test instance lifecycle is per-method, and as an instance-level extension if the lifecycle is per-class.

So, yeah, I admit: the terminology we use can be rather confusing.

In summary, I suppose we could say that extensions are invoked at one of the following levels, where _class level_ and _instance level_ can be somewhat generically grouped under the _container level_.

1. class level
2. instance level
3. method level

What I was trying convey earlier is that we have traditionally not considered that a _container-level_ invocation of an extension would ever have knowledge of anything at the method level (like the test `Method`).

**Comment by PaulSh84** [comment]
@sbrannen 
Can you please show at least one simple implementation of 

> TestInstanceFactory.createTestInstance()

I am looking for ways how to create additional test class instances and add different parameter dependencies in runtime.

**Comment by sbrannen** [comment]
@PaulSh84, please ask your question on [Gitter](https://gitter.im/junit-team/junit5#) or [Stack Overflow](https://stackoverflow.com/questions/tagged/junit5).

**Comment by PaulSh84** [comment]
@sbrannen 
I did, please take a look [here](https://stackoverflow.com/questions/64071113/how-can-i-control-how-many-junit5-test-instances-are-going-to-be-generated-in-ru)

**Comment by marcphilipp** [comment]
Tentatively slated for 5.8 M1 solely for the purpose of _team discussion_.

**Comment by marcphilipp** [comment]
@seanf Is this use case still relevant for you?

**Comment by michaeldimchuk** [comment]
@marcphilipp I don't know if the team is open to re-evaluating this feature request, but here's an example scenario where something like this is useful: https://github.com/quarkusio/quarkus/issues/38336.

In a scenario where all fields of a test class are `final` and are injected through the constructor with a `PER_METHOD` lifecycle, a CDI container has to be launched before the test instance is created. However at times the annotations applied to a method may be needed to configure the CDI container itself before actually launching it. 

In the case of [Quarkus](https://quarkus.io/guides/getting-started-testing#testing-components), they allow `@TestConfigProperty` annotations to be applied at either the class or method level, to override configuration properties defined for a test case, and allow for the container to be tested with those properties.

Since neither the `TestInstancePreConstructCallback` or `TestInstanceFactory` have access to the test method to be eventually invoked, something like that wouldn't be possible, and injection in that scenario has to be done after the test instance is created, so would only be applicable to non-final fields.

Being able to access the test method to be executed during test instance creation would allow for something like that to be done, while relying purely on constructor injection and not leaving mutable fields around, which works a bit better when using Lombok to apply `private final` to all non-static fields across a project.

**Comment by JojOatXGME** [comment]
FYI, I created a PR at #4032. The PR was written for #3445, but it would resolve this issue as well. However, since the change of the PR could potentially break existing extensions, more work may be required to provide proper legacy.

> > Extensions that implement `TestInstanceFactory` must be registered at the class level.
> 
> That means that the underlying `ExtensionContext` is scoped at the class level. Specifically, it is a `ClassExtensionContext` which has no access to the `Method`.

The Javadoc only declares where the extension has to be registered, it doesn't say anything about the `ExtensionContext` which is provided.

> Furthermore, doing so would not align with the design of JUnit Jupiter, namely that class-level extensions should not do anything particular to specific methods.

I am not sure about that. Extension callbacks are usually executed with the `ExtensionContext` matching the current state of the execution. For example, `ParameterResolver` will receive different `ExtensionContext` implementations depending on when a parameter needs to be resolved. As far as I know, the instance creation (with all the related extensions callbacks) is the only place which breaks this rule. I would therefore argue that the current behavior of `TestInstanceFactory` does not align with the design of JUnit Jupiter, as it behaves inconsistently to other extension callbacks. This difference also causes issues with `CloseableResource` in these extension methods, which is the topic of #3445.

Anyway, that is my perspective on this topic. Feel free to review my pull request. I would say my PR actually reduces the complexity of the production code a bit. However, there is still the problem that it might break existing extensions.

**Comment by marcphilipp** [comment]
Please see https://github.com/junit-team/junit5/issues/3445#issuecomment-2387853283.
